tests/test_basic.py

```diff
@@ -0,0 +1,91 @@
+import unittest
+
+from patchwork.actions import AskSubquestion, Reply, Unlock
+from patchwork.datastore import Datastore
+from patchwork.scheduling import RootQuestionSession, Scheduler
+
+
+# Note: This test is hard to read and tightly coupled with the implementation.
+# Over the next week I will make refactorings and implement infrastructure to
+# make it easy to write better tests. Or write more fine-grained unit tests
+# instead. I haven't decided yet. (RM 2018-06-09)
+class TestBasic(unittest.TestCase):
+    """Integration tests for basic scenarios."""
+
+    def testRecursion(self):
+        """Test the recursion example from the taxonomy.
+
+        Cf. https://ought.org/projects/factored-cognition/taxonomy#recursion
+        """
+        db      = Datastore()
+        sched   = Scheduler(db)
```

Comment:
I'd prefer not to align assignments, [following PEP8](https://www.python.org/dev/peps/pep-0008/#pet-peeves).

Reply:
What is your reason for that? PEP8 doesn't give any.

Reply:
The reason people usually give is that alignment makes diffs messy (because adding another variable with a longer name than the previous ones affects the previous one). But I mainly want the code to be formatted uniformly, and PEP8 seems like the best Schelling point. We might switch to auto-formatting code in the future, maybe using [Black](https://github.com/ambv/black), and auto-formatters usually follow PEP8 as well.

Reply:
I agree that disputes about code style are annoying and that official style guides and auto-formatters are a way to avoid them. In that spirit, this will be my last argument, after which I will submit to your judgment.

PEP8 says:

> Some other good reasons to ignore a particular guideline:
>
> 1. When applying the guideline would make the code less readable, even for someone who is used to reading code that follows this PEP.
> […]

Not aligning code vertically makes it less readable. PEP8 itself uses alignment in their examples:

```
     ↓
Yes: foo = (0,)
No:  bar = (0, )
```

Or take an example from Patchwork. How quickly can you understand what this does?

```
question_link = insert_raw_hypertext(contents, self.db, {})
answer_link = self.db.make_promise()
final_workspace_link = self.db.make_promise()
scratchpad_link = insert_raw_hypertext("", self.db, {})
new_workspace = Workspace(question_link, answer_link, final_workspace_link, scratchpad_link, [])
new_workspace_link = self.db.insert(new_workspace)
result = Context(new_workspace_link, self.db)
answer_link = self.db.dereference(result.workspace_link).answer_promise
self.active_contexts.add(result)
```

How quickly can you understand it now?

```
question_link           = insert_raw_hypertext(contents, self.db, {})
answer_link             = self.db.make_promise()
final_workspace_link    = self.db.make_promise()
scratchpad_link         = insert_raw_hypertext("", self.db, {})
new_workspace           = Workspace(question_link,
                                    answer_link,
                                    final_workspace_link,
                                    scratchpad_link,
                                    [])
new_workspace_link      = self.db.insert(new_workspace)

result                  = Context(new_workspace_link, self.db)
answer_link             = self.db.dereference(
                                result.workspace_link).answer_promise
self.active_contexts.add(result)
```

There seem to be more spaces than necessary in front of the equals signs. This is because I align [Go-style](https://golang.org/doc/effective_go.html#formatting), on multiples of four characters. This reduces the likelihood of messy diffs. In my experience this is not an issue anyway, because diff tools make it clear, it seldom occurs, and we read more code in the editor than in diffs.

Note that Go's autoformatter does align code.

I don't force other people to align their code, but I get sad when I'm less productive, because my whitespace is taken from me.